mirgecom/initializers.py

```diff
@@ -370,6 +465,76 @@ def exact_rhs(self, discr, q, t=0.0):
         return flat_obj_array(massrhs, energyrhs, momrhs)
 
 
+class AcousticPulse:
+    r"""
+    Create an N-dimensional Gaussian acoustic pulse.
+
+    The Gaussian pulse is defined by:
+
+    .. math::
+
+        {\rho}E(\vec{r}) = {\rho}E + amplitude * G(\vec{r})\\
+        G(\vec{r}) = \exp^{-(\frac{(\vec{r}-\vec{r_0})}{\sqrt{2}w})^{2}}
+
+    Where :math:`\vec{r}` are the nodal coordinates, and :math:`\vec{r_0}`,
+    :math:`amplitude`, and :math:`w`, are the the user-specified pulse location,
+    amplitude, and width, respectively.
+
+    .. automethod:: __init__
+    .. automethod:: __call__
+    """
+
+    def __init__(self, numdim=1, amplitude=1,
+                 center=None, width=1):
+        r"""
+        Initialize acoustic pulse parameters.
+
+        Parameters
+        ----------
+        numdim : int
+            specify the number of dimensions for the pulse
+        amplitude : float
+            specifies the value of :math:`amplitude`
+        width: float
+            specifies the rms width of the pulse
+        center : numpy.ndarray
+            pulse location, shape ``(numdim,)``
+        """
+        if len(center) == numdim:
+            self._center = center
+        elif len(center) > numdim:
+            numdim = len(center)
+            self._center = center
+        else:
+            self._center = np.zeros(shape=(numdim,))
+
+        assert len(self._center) == numdim
+
+        self._amp = amplitude
+        self._width = width
+        self._dim = numdim
+
+    def __call__(self, x_vec, q, eos=IdealSingleGas()):
+        """
+        Create the acoustic pulse at locations *x_vec*.
+
+        Parameters
+        ----------
+        t: float
+            Current time at which the solution is desired (unused)
+        x_vec: numpy.ndarray
+            Nodal coordinates
+        eos: :class:`mirgecom.eos.GasEOS`
+            Equation of state class to be used in construction of soln (unused)
+        """
+        assert len(x_vec) == self._dim
+        cv = split_conserved(self._dim, q)
+        cv.energy = cv.energy + _make_pulse(amp=self._amp, w=self._width,
+                                            r0=self._center, r=x_vec)
```

Comment:
These structures should not be assigned. Use `dataclasses.replace` instead to make a new one. We should probably declare them `frozen`.

Reply:
Oh shoot. That may spell a lot of trouble (for me).

Reply:
Turns out no too terribly much trouble.  I'm glad you caught this.

Reply:
> frozen

#135